bin/node/src/commands.rs

```diff
@@ -0,0 +1,168 @@
+use aleph_primitives::AuthorityId as AlephId;
+use log::info;
+use sc_cli::{CliConfiguration, Error, KeystoreParams, SharedParams};
+use sc_keystore::LocalKeystore;
+use sc_service::config::{BasePath, KeystoreConfig};
+use sp_application_crypto::key_types;
+use sp_consensus_aura::sr25519::AuthorityId as AuraId;
+use sp_core::crypto::KeyTypeId;
+use sp_core::{ed25519, sr25519};
+use sp_keystore::SyncCryptoStore;
+use std::io::Write;
+use std::sync::Arc;
+use structopt::StructOpt;
+
+/// The `bootstrap-chain` command is used generate keys for the genesis authorities
+/// keys are written to the keystore of the authorities
+/// and the chain specification is printed to stdout in the JSON format
+#[derive(Debug, StructOpt)]
+pub struct BootstrapChainCmd {
+    /// Force raw genesis storage output.
+    #[structopt(long = "raw")]
+    pub raw: bool,
+
+    #[structopt(flatten)]
+    pub keystore_params: KeystoreParams,
+
+    #[structopt(flatten)]
+    pub shared_params: SharedParams,
+}
+
+impl CliConfiguration for BootstrapChainCmd {
+    fn shared_params(&self) -> &SharedParams {
+        &self.shared_params
+    }
+}
+
+impl BootstrapChainCmd {
+    fn upsert_aura_key(
+        &self,
+        keystore: &LocalKeystore,
+        key_type: KeyTypeId,
+    ) -> Result<AuraId, Error> {
+        let key = SyncCryptoStore::sr25519_public_keys(&*keystore, key_type)
+            .into_iter()
+            .next()
+            .map_or_else(
+                || {
+                    SyncCryptoStore::sr25519_generate_new(&*keystore, key_type, None)
+                        .map_err(|_| Error::KeyStoreOperation)
+                },
+                Ok,
+            );
+
+        match key {
+            Ok(k) => {
+                let bytes = k.as_array_ref();
+                Ok(AuraId::from(sr25519::Public::from_raw(*bytes)))
+            }
+            Err(why) => panic!("{}", why),
+        }
+    }
+
+    fn upsert_aleph_key(
+        &self,
+        keystore: &LocalKeystore,
+        key_type: KeyTypeId,
+    ) -> Result<AlephId, Error> {
+        let key = SyncCryptoStore::ed25519_public_keys(&*keystore, key_type)
+            .into_iter()
+            .next()
+            .map_or_else(
+                || {
+                    SyncCryptoStore::ed25519_generate_new(&*keystore, key_type, None)
+                        .map_err(|_| Error::KeyStoreOperation)
+                },
+                Ok,
+            );
+
+        match key {
+            Ok(k) => {
+                let bytes = k.as_array_ref();
+                Ok(AlephId::from(ed25519::Public::from_raw(*bytes)))
+            }
+            Err(why) => panic!("{}", why),
+        }
+    }
+
+    pub fn run(&self, cli: &Cli) -> Result<(), Error> {
+        let chain_id = self.shared_params.chain_id(self.shared_params.is_dev());
+
+        let mut genesis_authorities: Vec<AuthorityKeys> = Vec::new();
+
+        LOCAL_AUTHORITIES.iter().for_each(|authority| {
```

Comment:
calling `.map(|authority| {...}).collect()` instead of `for_each(...)` would be cleaner and would allow us to get rid of the ugly `mut` in `genesis_authorities`. Also, `for` loop is more idiomatic than `for_each`.

Reply:
good spot, a lot more functional, thanks :)